core/node/track_streams/multi_sync_runner.go

```diff
@@ -0,0 +1,776 @@
+	// These are used to rate limit concurrent requests to each remote node from this service.
+	concurrentNodeRequests sync.Map // map[commonAddress]*semaphore.Weighted
+
+	// unfilledSyncs tracks sync sessions that have not yet been filled to maximum capacity for each
+	// node. Once a sync session unassignable, we remove the reference here. The sync session continues
+	// to be accessed via it's go routine, but it will be garbage collected if the routine exits.
+	// In such cases the sync session runner will first relocate all of its streams.
+	unfilledSyncs sync.Map // map[commonAddress]*syncStreamRunner
+
+	// TODO: use a single node registry and modify http client settings for better performance.
+	nodeRegistries []nodes.NodeRegistry
+}
+
+// getNodeRequestPool returns the node-specific semaphore used to rate limit requests to each node
+// on the network.
+func (msr *MultiSyncRunner) getNodeRequestPool(addr common.Address) *semaphore.Weighted {
+	if workerPool, ok := msr.concurrentNodeRequests.Load(addr); ok {
+		return workerPool.(*semaphore.Weighted)
+	}
+
+	workerPool, _ := msr.concurrentNodeRequests.LoadOrStore(
+		addr,
+		semaphore.NewWeighted(int64(msr.config.MaxConcurrentNodeRequests)),
+	)
+	return workerPool.(*semaphore.Weighted)
+}
+
+// NewMultiSyncRunner creates a MultiSyncRunner instance.
+func NewMultiSyncRunner(
+	metrics *TrackStreamsSyncMetrics,
+	onChainConfig crypto.OnChainConfiguration,
+	nodeRegistries []nodes.NodeRegistry,
+	trackedStreamViewConstructor TrackedViewConstructorFn,
+	streamTrackingConfig config.StreamTrackingConfig,
+	otelTracer trace.Tracer,
+) *MultiSyncRunner {
+	// Set configuration defaults if needed
+	if streamTrackingConfig.NumWorkers < 1 {
+		streamTrackingConfig.NumWorkers = 20
+	}
+	if streamTrackingConfig.StreamsPerSyncSession < 1 {
+		streamTrackingConfig.StreamsPerSyncSession = 100
+	}
+	if streamTrackingConfig.MaxConcurrentNodeRequests < 1 {
+		streamTrackingConfig.MaxConcurrentNodeRequests = 50
+	}
+
+	return &MultiSyncRunner{
+		metrics:                metrics,
+		onChainConfig:          onChainConfig,
+		nodeRegistries:         nodeRegistries,
+		trackedViewConstructor: trackedStreamViewConstructor,
+		streamsToSync:          make(chan (streamSyncInitRecord), 2048),
+		config:                 streamTrackingConfig,
+		workerPool:             *workerpool.New(streamTrackingConfig.NumWorkers),
+		otelTracer:             otelTracer,
+	}
+}
+
+// Run starts the operation of the MultiSyncRunner and continues to add streams to sync sessions until rootCtx is canceled.
+func (msr *MultiSyncRunner) Run(
+	rootCtx context.Context,
+) {
+	for range msr.config.NumWorkers {
+		// Each worker routine will continuously loop, placing streams in sync sessions, until the rootCtx is canceled.
+		msr.workerPool.Submit(func() {
+			for {
+				select {
+				case <-rootCtx.Done():
+					return
+				case streamInitRecord := <-msr.streamsToSync:
+					msr.addToSync(rootCtx, streamInitRecord)
+				}
+			}
+		})
+	}
```

Comment:
Instead of having `msr.config.NumWorkers` all waiting on `msr.streamsToSync` is it possible to have the go-routine that called Run listing from `msr.streamsToSync` and post add to sync requests to the worker pool? This approach lets the workerpool manage go-routines instead of always having NumWorkers running.